### PR TITLE
JIT: Fix weights in fgOptimizeUncondBranchToSimpleCond

### LINKED ISSUE #50419: JIT: PGO: fgOptimizeUncondBranchToSimpleCond corrupts profile data
It looks like `fgOptimizeUncondBranchToSimpleCond` (as part of the `Optimize Layouts` phase) produces invalid weights in some cases. E.g.

Before `Optimize layout`:
image

^ all edges look valid and satisfy Kirchhoff's Law.

After `Optimize Layouts`:

image

^ 
invalid edges:
BB06 (in: 0 + 0.5, out: 1)
BB09 (in: 0.25, out: 0)

I guess invalid edges are fine and will be re-calculated, but there are also invalid blocks such as BB09 which is a predecessor of a cold block but has weight 0.5 (BB05 should also be cold then?)

As the result, in the end I have:
```
; with IBC profile data, edge weights are invalid, and fgCalledCount is 36
```
But if I disable `fgOptimizeUncondBranchToSimpleCond` the final weights will be reported as valid:

image

^ "After Optimize Layout" without `fgOptimizeUncondBranchToSimpleCond`, all blocks look valid again (edges aren't though).

/cc @AndyAyersMS 

category:correctness
theme:profile-feedback
skill-level:beginner
cost:small
impact:small

### PR DESCRIPTION
**Pull request by EgorBo**
Fixes https://github.com/dotnet/runtime/issues/50419

Repro (`COMPlus_TieredPGO=1`):
```csharp
using System.Runtime.CompilerServices;
using System.Threading;

class Program
{
    static void Main()
    {
        // Collect some profile, wait for tier0->tier1 promotion.
        for (int i = 0; i < 100; i++)
        {
            Test(new object());
            Thread.Sleep(16);
        }
    }

    static object someObj;

    [MethodImpl(MethodImplOptions.NoInlining)]
    static void Test(object o)
    {
        if (o != someObj)
            o = null;

        if (o != null)
            Test1();
    }

    [MethodImpl(MethodImplOptions.NoInlining)] static void Test1() { }
}
```
Codegen diff between main and this PR: https://www.diffchecker.com/iqSxlzMG (note "edge weights are invalid" on the left)

image

The edges' weights aren't even rendered on the left (same phase).

### CODE REVIEW

**Comment by EgorBo** [comment]
I've tested this on a heavy desktop app (https://github.com/icsharpcode/AvaloniaILSpy)

Before: 367 methods are reported to have invalid weights
After: 338 

(investigating other failures)

**Comment by EgorBo** [comment]
PTAL @AndyAyersMS @dotnet/jit-contrib

**Review by AndyAyersMS — COMMENTED** [review]
I  think it needs to be more general than this. Try some cases where the condition is not always true or not always false.

We have something like this to begin with. Say we know all the block and edge weights and they are consistent and weight(X) is not zero (if it is we perhaps should not do this opt):

image - 2021-03-31T132808 350

We want to duplicate X into A (expecting that the one of the paths out of A will get optimized away):

image - 2021-03-31T133131 051

So the block weight of X needs to decrease (it was A+B, it should now be just B). The block weight of T is unknowable -- we have to make a guess here. There is no good answer; the simplest is to give the outgoing edges of A the same weight ratio as the outgoing edges of X. Then we're consistent if we don't optimize, but inconsistent if we do; that's ok as it means we've learned something contextual. We'll clean such things up in some future work items.

So the updates should be something like:
```
ScaleA = weight (A)  / weight(X)
ScaleB = weight (B)  / weight(X)   ;; or 1-ScaleA

weight(X->C) = weight(X->C) * ScaleB
weight(X->D) = weight(X->D) * ScaleB
weight(A->C) = weight(X->C) * ScaleA
weight(A->T)  = weight(X->D) * ScaleA
weight(T) = weight(X->D) * ScaleA
weight(X) = weight(B)
```

**Comment by AndyAyersMS** [comment]
Since there can actually be many other preds, a cleaner expression is perhaps:
```
ScaleA = weight(A) / weight(X)
ScaleNotA = 1 - ScaleA

weight(X)    = weight(X)    * ScaleNotA
weight(X->C) = weight(X->C) * ScaleNotA
weight(X->D) = weight(X->D) * ScaleNotA

weight(A->C) = weight(X->C) * ScaleA
weight(A->T) = weight(X->D) * ScaleA
weight(T)    = weight(X->D) * ScaleA
```

**Comment by EgorBo** [comment]
Ping myself

**Comment by EgorBo** [comment]
@AndyAyersMS could you please take a look, I've integrated your changes

File: src/coreclr/jit/fgopt.cpp [hunk 1]
```diff
@@ -3208,6 +3208,16 @@ bool Compiler::fgOptimizeUncondBranchToSimpleCond(BasicBlock* block, BasicBlock*
...
     assert(!block->IsLIR());
 
+    // See https://github.com/dotnet/runtime/pull/50490#pullrequestreview-625700642
+    //
+    // A - block
+    // X - target
+    // C - target->bbJumpDest
+    // D - target->bbNext
+
+    const BasicBlock::weight_t scaleA    = block->bbWeight / target->bbWeight;
```
Comment (AndyAyersMS) [inline src/coreclr/jit/fgopt.cpp:3218]: You probably should watch for division by zero here.

**Review by AndyAyersMS — COMMENTED** [review]
Changes LGTM, but there's an assert in crossgen2 to investigate.

File: src/coreclr/jit/fgopt.cpp [hunk 2]
```diff
@@ -3207,6 +3207,22 @@ bool Compiler::fgOptimizeUncondBranchToSimpleCond(BasicBlock* block, BasicBlock*
     // `fgUpdateFlowGraph` has been called with `doTailDuplication` set to true, and the
     // backend always calls `fgUpdateFlowGraph` with `doTailDuplication` set to false.
     assert(!block->IsLIR());
+    
+    // The optimization doesn't make sense in this case
+    if (target->bbWeight == 0.0)
+    {
+        return false;
+    }
```
Comment (AndyAyersMS) [inline src/coreclr/jit/fgopt.cpp:3215]:
```suggestion
    // The optimization doesn't make sense in this case
    if (target->isRunRarely())
    {
        return false;
    }
```
Also this check is quite cheap so I'd move it up to the top of the method as the first check.

**Comment by EgorBo** [comment]
Ping myself

**Pull request closed without merging**